modules/core/src/main/java/org/apache/fluo/core/impl/AsyncReader.java

```diff
@@ -0,0 +1,100 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one or more contributor license
+ * agreements. See the NOTICE file distributed with this work for additional information regarding
+ * copyright ownership. The ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the License. You may obtain a
+ * copy of the License at
+ * 
+ * http://www.apache.org/licenses/LICENSE-2.0
+ * 
+ * Unless required by applicable law or agreed to in writing, software distributed under the License
+ * is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package org.apache.fluo.core.impl;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.BlockingQueue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.LinkedBlockingQueue;
+
+import com.google.common.collect.Collections2;
+import org.apache.fluo.api.data.Bytes;
+import org.apache.fluo.api.data.Column;
+import org.apache.fluo.api.data.RowColumn;
+import org.apache.fluo.core.impl.TransactionImpl;
+
+public class AsyncReader {
+  private BlockingQueue<AsyncGet> asyncGetsQueue;
+  private ExecutorService executorService;
+  private TransactionImpl tx;
+
+  public AsyncReader(TransactionImpl tx) {
+    this.tx = tx;
+    asyncGetsQueue = new LinkedBlockingQueue<>();
+    executorService = Executors.newSingleThreadExecutor();
+  }
+
+  public CompletableFuture<Bytes> get(Bytes row, Column column) {
+    return get(row, column, null);
+  }
+
+  public CompletableFuture<Bytes> get(Bytes row, Column column, Bytes defaultValue) {
+    AsyncGet curAsyncGet = new AsyncGet(row, column, defaultValue);
+    asyncGetsQueue.add(curAsyncGet);
+
+    executorService.submit(() -> {
+      List<AsyncGet> getsList = new ArrayList<>();
+      asyncGetsQueue.drainTo(getsList);
+
+      try {
+        Collection<RowColumn> rowColumns = Collections2.transform(getsList, ag -> ag.rc);
+
+        Map<RowColumn, Bytes> getsMap = tx.get(rowColumns);
+
+        for (AsyncGet asyncGet : getsList) {
+          Bytes result = getsMap.get(asyncGet.rc);
+          asyncGet.res.complete(result == null ? defaultValue : result);
+        }
+      } catch (Exception e) {
+        for (AsyncGet asyncGet : getsList) {
+          asyncGet.res.completeExceptionally(e);
```

Comment:
If an exception were to occur in the for loop inside the try, then it is possible that some of the futures are already completed.  That case seems unexpected, not sure what if anything we should do about that case.

Reply:
How about removing each AsyncGet object from getsList as we complete it?

Reply:
I was looking at the [javadoc](https://docs.oracle.com/javase/8/docs/api/java/util/concurrent/CompletableFuture.html#completeExceptionally-java.lang.Throwable-) and I don't think anything needs to be done.  The javadoc for completeExceptionally says the following

```
If not already completed, causes invocations of get() and related methods to throw the given exception.
```

So it seems if its already completed, then nothing will happen.